draft-ietf-mls-protocol.md

```diff
@@ -1008,14 +1112,178 @@ struct {
 } UserInitKey;
 ~~~~~
 
+# Message Framing
+
+Handshake and application messages use a common framing structure.
+This framing provides encryption to assure confidentiality within the
+group, as well as signing to authenticate the sender within the group.
+
+The two main structures involved are MLSPlaintext and MLSCiphertext.
+MLSCiphertext represents a signed and encrypted message, with
+protections for both the content of the message and related
+metadata.  MLSPlaintext represents a message that is only signed,
+and not encrypted.  Applications SHOULD use MLSCiphertext to encode
+both application and handshake messages, but MAY transmit handshake
+messages encoded as MLSPlaintext objects in cases where it is
+necessary for the delivery service to examine such messages.
+
+~~~~~
+enum {
+    invalid(0),
+    handshake(1),
+    application(2),
+    (255)
+} ContentType;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    ContentType content_type;
+
+    select (MLSPlaintext.type) {
+        case handshake:
+            GroupOperation operation;
+
+        case application:
+            opaque application_data<0..2^32-1>;
+    }
+
+    opaque signature<0..2^16-1>;
+} MLSPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    ContentType content_type;
+    opaque masked_sender_data[8];
+    opaque ciphertext<0..2^32-1>;
+} MLSCiphertext;
+~~~~~
+
+The remainder of this section describe how to compute the signature of
+an MLSPlaintext object and how to convert it to an MLSCiphertext object.
+The overall process is as follows:
+
+* Compute the metadata by concatenating the following information:
```

Comment:
"Concatenating" is not quite right, since it's a bit more complex than that.  Maybe "Gather the required metadata:"